text/0321-key-functions.md

```diff
@@ -0,0 +1,74 @@
+- Start Date: 2018-03-24
+- RFC PR: https://github.com/emberjs/rfcs/pull/321
+- Ember Issue: (leave this empty)
+
+# Key Functions for `each`/`each-in`
+
+## Summary
+
+This RFC proposes that we support passing functions to the `key` argument of the `each` and `each-in` helpers. These functions would be called for each item and would be expected to return they key value used in diffing.
+
+For example,
+
+```js
+import Component from '@ember/component';
+
+export default Component.extend({
+  itemKey(item) {
+    return `${item.factoryNumber}-${item.partNumber}`;
+  }
+});
+```
+
+```hbs
+{{#each @items key=itemKey as |item|}}
+  {{item-details item=item}}
+{{/each}}
+```
+
+## Motivation
+
+The `key` argument to `each` and `each-in` determines the key used in the Glimmer diffing algorithm. Here is a description of the currently accepted values?
+
+`key` argument          | Computed key
+------------------------|---------------------------------------------
+`@identity`             | `guidFor(item)`
+`@index`                | index of the item in the array / object keys
+`@key` (`each-in` only) | the key of the key-value pair
+other strings           | `get(item, path)`
+otherwise               | error
+
+If a key is not specified then `@identity` strategy is used.
+
+Unfortunately, unless one of the default strategies works for your use case, you're stucking using a path on your object as the key. This can be problematic when you need to assemble a composite key from multiple values as in the example above. It is especially problematic when dealing with objects whose fields are out of your control (e.g. a moment date, an immutable.js record, a frozen object, etc.). Another less common reason you might want to use a key function is if your keys are expensive to compute and you want to cache them in a `WeakMap`.
+
+## Detailed design
+
+If `key` is a function, then Glimmer's diffing algorithm will call this function to determine the item's key.
+
+For `each`, the function would be called with arguments `keyFn(item, index)`.
+For `each-in`, the function would be called with arguments `keyFn(key, value)`.
+
+The key function should not be invoked with a `this`.
+
+One additional benefit of key functions is that they can be used to model the existing strategies in a uniform way:
+
+`key` argument | `each` key function                | `each-in` key function
+---------------|------------------------------------|-----------------------------------
+`@identity`    | `(item, index) => guidFor(item)`   | `(key, value) => guidFor(value)`
```

Comment:
One possible suggestion I have is to reverse the order of the arguments in one of the functions. I believe that the value should be in the same position in both.

`key` argument | `each` key function                | `each-in` key function
---------------|------------------------------------|-----------------------------------
`@identity`    | `(item, index) => guidFor(item)`   | `(value, key) => guidFor(value)`

or

`key` argument | `each` key function                | `each-in` key function
---------------|------------------------------------|-----------------------------------
`@identity`    | `(index, item) => guidFor(item)`   | `(key, value) => guidFor(value)`

I have the feeling that it would make the table simpler:

`key` argument |  key function for both
---------------|------------------------------------
`@identity`    | `(_, value) => guidFor(value)`  
`@index`/`@key`       | `(ref, _) => ref`   
other strings  | `(_, value) => get(value, path)`

However I also see the convenience on `{{each}}` taking the arguments in the same order that `.forEach` and `{{each-in}}` having them as (`key`, `value`)

Reply:
That order seems plausible since it's the order you get when calling the Array.prototype.entries and Map.prototype.entries iterators, however I think it still makes more sense to mirror the block param order. Additionally, I don't like promoting array index to the first slot because you almost never want to use it as a key (unless you know the whole #each block is state-less and no one will ever add a stateful component into it in the future, see https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318).

Reply:
Yeah, I'm not really opposed to the proposed order, I just thought that this one presented a "chance for unification", in case that was desirable or simplified the implementation in any form. I'm happy either way.